core/base/src/main/java/org/eclipse/dataspaceconnector/core/executor/ExecutorInstrumentationImplementation.java

```diff
@@ -0,0 +1,30 @@
+/*
+ *  Copyright (c) 2022 Microsoft Corporation
+ *
+ *  This program and the accompanying materials are made available under the
+ *  terms of the Apache License, Version 2.0 which is available at
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Contributors:
+ *       Microsoft Corporation - initial API and implementation
+ *
+ */
+
+package org.eclipse.dataspaceconnector.core.executor;
+
+import org.eclipse.dataspaceconnector.spi.system.ExecutorInstrumentation;
+
+/**
+ * Interface for implementations of {@link ExecutorInstrumentation}s.
+ * <p>
+ * A distinct interface is provided in order for the dependency injection
+ * to provide a default implementation if no other implementation is
+ * provided in the classpath.
+ * <p>
+ * The default implementation does not provide any instrumentation. Extension
+ * modules can provide implementations, such as for collecting metrics.
+ */
+public interface ExecutorInstrumentationImplementation extends ExecutorInstrumentation {
```

Comment:
Actually both interfaces are needed. `ExecutorInstrumentation` is the main interface used to instrument the `ExecutorService`. The `CoreServicesExtension` contains the logic to load the corresponding implementation by finding implementations of `ExecutorInstrumentationImplementation` like for instance the one provided in the `MicrometerExtension`. In case no implementation is available then the default `NoopExecutorInstrumentationImplementation` is used.

Another way to see it is:
- `CoreServicesExtension` takes care of `@Provides({ExecutorInstrumentation.class})'
- `MicrometerExtension` takes care of `@Provides({ExecutorInstrumentationImplementation.class})'
having only a single interface would result in a duplicate `@Provides`.

I know it might be confusing and I am open for suggestions how to improve it.

Reply:
Ok understood, but could we achieve the same goal with the following sequence:
- The `CoreServicesExtension` registers a `NoopExecutorInstrumentationImpl`
- The `MicrometerExtension` registers the `MicrometerExecutorInstrumentation`, effectively overwriting the `Noop...` one, since `CoreServicesExtension` gets initialized first.

This "provide-default-and-overwrite" pattern is used on several other locations in the code.

Reply:
I see. That pattern actually surprised me because I would expect an IoC container to either take the first implementation (to honour classpath order) or crash if > 1 bean match. Another thing worth documenting.

Reply:
In all honesty it is not (yet?) a fully-fledged DI container, in order to become one it misses a few features, such as collection injection and injection by qualifier ("named injections"). This "default-and-overwrite" pattern is actually pretty useful, because it allows us to provide default implementations, such as in-memory stores.

Reply:
@paullatzelsperger Assuming we get rid of the `ExecutorInstrumentationImplementation` interface, the problem with your suggestion is that the `CoreServicesExtension` itself requires an `ExecutorInstrumentation` to feed it into the `HealthCheckService`, so actually I think it would be the `MicrometerExtension` being loaded first (I have actually verified this by debugging).

An alternative would be to have `CoreServicesExtension` load the noop implementation only if no other one could be loaded before.

Reply:
> ... so actually I think it would be the `MicrometerExtension` being loaded first (I have actually verified this by debugging).
> 
> An alternative would be to have `CoreServicesExtension` load the noop implementation only if no other one could be loaded before. I have now implemented it this way, have a look!

sounds good, will take a look!